### ISSUE TITLE
sysutils/nut [os-nut]: upsdrvctl shutdown does not get triggered

### ISSUE THREAD

**Issue by spessoni** [opnsense/plugins#3013]
**Important notices**
Before you add a new report, we ask you kindly to acknowledge the following:

- [x] I have read the contributing guide lines at https://github.com/opnsense/plugins/blob/master/CONTRIBUTING.md
- [x] I have searched the existing issues, open and closed, and I'm convinced that mine is new.
- [x] The title contains the plugin to which this issue belongs

**Describe the bug**
`upsdrvctl shutdown` does not get triggered. This could lead to a state in which the UPS remains on but OPNsense is powered down with no way to recover.

**To Reproduce**

1. Go to 'NUT --> Configuration --> General Settings'
2. Check 'Enable Nut'
3. Switch 'Service Mode' to 'standalone'
4. Set 'Name' to 'ups'
5. Click on 'Apply'
6. Go to 'NUT --> Configuration --> UPS Type --> USBHID-Driver'
7. Check 'enable'
8. Click on 'Apply'
9. Execute a forced shutdown via `upsmon -c fsd`
10. OPNsense powers off, but the UPS remains running

**Expected behavior**
`upsdrvctl shutdown` should be triggered during a upsmon forced shutdown event.  Executing `upsdrvctl shutdown` during an event will start the UPS shutdown timer, which will turn the UPS off and wait for power to return. If power returns during a shutdown, the UPS will power cycle (turn off for about a second, and turn back on) and force the server to startup (if set in the BIOS).

**Additional context**
I believe there is actually 2 problems here:

1. The `upsmon -K` test does not work. It is always exits with `EXIT_FAILURE` regardless if `/etc/killpower` is present or not. Replacing this with `test -f /etc/killpower` does work.

2. `nut_upsshut="YES"` is never set in `/etc/rc.conf.d/nut`.  Because this is never set, the test in `/usr/local/etc/rc.d/nut` fails.

I haven't investigated why `upsmon -K` does not work, however `test -f /etc/killpower` does. Perhaps permissions?

I believe adding a checkbox in the UI to _enable UPS Shutdown_, which updates the template to add `nut_upsshut="YES"` would correct this. By default, in `/usr/local/etc/rc.d/nut`, it is set to NO.

**Reference `/usr/local/etc/rc.d/nut`:**
This is where the UPS power off test are. Im not sure where this is in the plugin source code.

Top of file. Notice default is set to NO: `nut_upsshut=${nut_upsshut:-"NO"}`
```
#!/bin/sh

# PROVIDE: nut
# REQUIRE: NETWORKING devfs
# BEFORE: LOGIN
# KEYWORD: shutdown

# Define these nut_* variables in one of these files:
#       /etc/rc.conf
#       /etc/rc.conf.local
#       /etc/rc.conf.d/nut
#
# DO NOT CHANGE THESE DEFAULT VALUES HERE
#
nut_enable=${nut_enable:-"NO"}
nut_prefix=${nut_prefix:-"/usr/local"}
nut_upsshut=${nut_upsshut:-"NO"}

```

The **stop** function:
```
nut_poststop() {
	if ${nut_prefix}/sbin/upsdrvctl stop && checkyesno nut_upsshut; then
		if ${nut_prefix}/sbin/upsmon -K; then
			${nut_prefix}/sbin/upsdrvctl shutdown
		fi
	fi

}
```

**Template for `/etc/rc.conf.d/nut`:**

https://github.com/opnsense/plugins/blob/285dcd88c3a99f1d38c4f7ad1d8a4546daa08170/sysutils/nut/src/opnsense/service/templates/OPNsense/Nut/nut#L1-L6

**Temporary Workaround / Proof of Concept**
I created a file called `99-nut` in `/usr/local/etc/rc.syshook.d/stop` and made it executable. I added some beeps so I could hear what mode was detected and executed.

```
#!/bin/sh

# This does NOT work:
# (/usr/local/sbin/upsmon -K)

# This does work:
# (test -f /etc/killpower)

if (test -f /etc/killpower); then
	echo "UPS: Power Failure Shutdown - UPS Shutdown Activated"
	opnsense-beep high
	opnsense-beep high
	opnsense-beep low
	opnsense-beep high
	opnsense-beep high
	/usr/local/sbin/upsdrvctl shutdown
else
	echo "UPS: Standard Shutdown - No UPS shutdown"
	opnsense-beep low
	opnsense-beep low
	opnsense-beep high
	opnsense-beep low
	opnsense-beep low
fi
```

**Helpful Links**

- This was helpful in understanding FreeBSD / NUT during a Low Battery Shutdown Procedure:
https://blog.tyk.nu/blog/going-nuts-using-a-ups-on-freebsd/
- Similar Issue: [NUT: no restart of UPS](https://forum.opnsense.org/index.php?topic=23164.msg)
https://forum.opnsense.org/index.php?topic=23164.0

**Environment**
OPNsense 22.1.8_1 (amd64)
os-nut 1.8.1

**Comment by FudiHub** [comment]
@spessoni, Thanks for the workaround you added.
I tried it and it seems to work pretty good.
Does the created 99-nut file survive a opnsense updates?

Hopefully the ups-shut switch is added to nut plugin soon.

**Comment by spessoni** [comment]
Yes! It does survive the reboot. Ideally the UPS shutdown should run after the filesystem is set to read only, as some UPS's don't support the delayed shutdown correctly.

I'm not sure if I should tag the OpnSense NUT maintainer on this, as this continues to be broken, or if that's the appropriate way to raise awareness.

**Comment by FudiHub** [comment]
Not sure as well. I really appreciate the details of your issue description. I figured out the same but I am not able to describe it like you. Good job, and thanks for that. The workaround helps me a lot as well.

**Comment by fichtner** [comment]
It looks a bit like nut_upsshut=YES would be required but the script is still broken (FreeBSD ports issue?)

I don't mind adding that syshook script to the plugin but doubting that 99 priority is appropriate. I'm sure maintainer @mimugmail would mind.

PRs welcome :)

**Comment by mimugmail** [comment]
It would be better to fix it in FreeBSD ports

**Comment by fichtner** [comment]
The plugin needs to be adjusted in any case 😉

**Comment by aptalca** [comment]
Can this issue be reopened in the meantime? I just found out about this the hard way. Power loss, opnsense shut down gracefully. UPS remained on and power came back before UPS depleted the battery. Opnsense remained off and I was left with no router and no access to homelab until I got home later.

Thanks

**Comment by spessoni** [comment]
I agree! 
How do we communicate this up to the FreeBSD port for this issue:

> 1. The `upsmon -K` test does not work. It is always exits with `EXIT_FAILURE` regardless if `/etc/killpower` is present or not. Replacing this with `test -f /etc/killpower` does work.

And what can we do to get this in the OpnSense plugin:
> 2\. `nut_upsshut="YES"` is never set in `/etc/rc.conf.d/nut`.  Because this is never set, the test in `/usr/local/etc/rc.d/nut` fails.
> 
> I believe adding a checkbox in the UI to _enable UPS Shutdown_, which updates the template to add `nut_upsshut="YES"` would correct this. By default, in `/usr/local/etc/rc.d/nut`, it is set to NO.

**Comment by glasid** [comment]
> It would be better to fix it in FreeBSD ports

There's nothing to fix in FreeBSD. Everything is working as intended.

The administrator must manually set `nut_upsshut="YES"` if UPS shutdown is desired.

In case of OPNsense the NUT plugin needs to be adjusted to implement the respective option.

**Comment by aptalca** [comment]
> There's nothing to fix in FreeBSD. Everything is working as intended.

Can you elaborate? The post above yours details the issue about the test always failing (not proper conditional). That's the freebsd issue. You can't just drop in with a `there is no issue` claim and no further details. 

> The administrator must manually set nut_upsshut="YES" if UPS shutdown is desired.

> In case of OPNsense the NUT plugin needs to be adjusted to implement the respective option.

In case you misunderstood, we were talking about two separate issues. The first one was a broken test, which was a freebsd issue. The second was the missing shutdown option in the opnsense plugin.

With that said, this github issue hasn't gotten any responses in a few years, so perhaps those issues are fixed now. I wouldn't know as I manually modified the nut plugin to get it to work then and later switched to apcupsd.

**Comment by spessoni** [comment]
I am still using the workaround with 99-nut. 
I'm sure the plugin still needs the nut_upsshut="YES" and haven't tested the "upsmon -K" lately.

**Comment by glasid** [comment]
> Can you elaborate? The post above yours details the issue about the test always failing (not proper conditional). That's the freebsd issue. You can't just drop in with a `there is no issue` claim and no further details.

This issue has been fixed long time ago with NUT version 2.8.1. Hence, the test doesn't fail.
[Commit a679e52](https://github.com/networkupstools/nut/commit/a679e521aabdaf694c5d96e2e2faa960bc8a653b)

```
		if (cmd) {
			upsdebugx(1, "Signaled old daemon OK");
		} else {
			if (checking_flag) {
				printf("Note: A previous upsmon instance is already running!\n");
				printf("Usually it should not be running during OS shutdown,\n");
				printf("which is when checking POWERDOWNFLAG makes most sense.\n");
			} else {
				printf("Fatal error: A previous upsmon instance is already running!\n");
				printf("Either stop the previous instance first, or use the 'reload' command.\n");
				exit(EXIT_FAILURE);
			}
		}
		break;
```

When you call upsmon -K , _checking_flag_ will be set to 1.


> 
> > The administrator must manually set nut_upsshut="YES" if UPS shutdown is desired.
> 
> > In case of OPNsense the NUT plugin needs to be adjusted to implement the respective option.
> 
> In case you misunderstood, we were talking about two separate issues. The first one was a broken test, which was a freebsd issue. The second was the missing shutdown option in the opnsense plugin.
> 

Without any shutdown option in the opnsense plugin the variable _nut_upsshut_ will always be NO.

`nut_upsshut=${nut_upsshut:-"NO"}`

Hence, the following if clause will be false and upsdrvctl shutdown command won't be called.

`if ${nut_prefix}/sbin/upsdrvctl stop && checkyesno nut_upsshut; then`

**Comment by Billy5804** [comment]
> I am still using the workaround with 99-nut. I'm sure the plugin still needs the nut_upsshut="YES" and haven't tested the "upsmon -K" lately.

`upsmon -K` is working now, so you could modify your 99-nut script.


> Hence, the following if clause will be false and upsdrvctl shutdown command won't be called.
<br>`if ${nut_prefix}/sbin/upsdrvctl stop && checkyesno nut_upsshut; then`

For some reason `upsdrvctl stop` half of that check errors on my setup, so I'd still have to use a workaround.
